From ee46c3bc11e667ec68eb3e992d0b7bd701157eff Mon Sep 17 00:00:00 2001
From: Pete Beardmore <pete.beardmore@msn.com>
Date: Sat, 17 Mar 2018 21:17:43 +0000
Subject: core, refactor 'ExpandBoxHack'

-move to common 'qltk.x' path
-rename to more appropriate 'ExpanderTitleContainerHack'
-add '__init__' function to populate content
---
 quodlibet/quodlibet/qltk/queue.py | 20 +++-----------------
 quodlibet/quodlibet/qltk/x.py     | 19 +++++++++++++++++++
 2 files changed, 22 insertions(+), 17 deletions(-)

diff --git a/quodlibet/quodlibet/qltk/queue.py b/quodlibet/quodlibet/qltk/queue.py
index 2352b977f..6d052f295 100644
--- a/quodlibet/quodlibet/qltk/queue.py
+++ b/quodlibet/quodlibet/qltk/queue.py
@@ -21,7 +21,7 @@ from quodlibet import qltk
 from quodlibet import app
 
 from quodlibet.util import connect_destroy, format_time_preferred, print_exc
-from quodlibet.qltk import Icons, gtk_version, add_css
+from quodlibet.qltk import Icons, add_css
 from quodlibet.qltk.ccb import ConfigCheckMenuItem
 from quodlibet.qltk.songlist import SongList, DND_QL, DND_URI_LIST
 from quodlibet.qltk.songsmenu import SongsMenu
@@ -29,7 +29,7 @@ from quodlibet.qltk.menubutton import SmallMenuButton
 from quodlibet.qltk.songmodel import PlaylistModel
 from quodlibet.qltk.playorder import OrderInOrder, OrderShuffle
 from quodlibet.qltk.x import ScrolledWindow, SymbolicIconImage, \
-    SmallImageButton, MenuItem
+    SmallImageButton, MenuItem, ExpanderTitleContainerHack
 
 QUEUE = os.path.join(quodlibet.get_user_dir(), "queue")
 
@@ -64,20 +64,6 @@ class PlaybackStatusIcon(Gtk.Box):
         self._set("media-playback-pause")
 
 
-class ExpandBoxHack(Gtk.HBox):
-
-    def do_get_preferred_width(self):
-        # Workaround for https://bugzilla.gnome.org/show_bug.cgi?id=765602
-        # set_label_fill() no longer works since 3.20. Fake a natural size
-        # which is larger than the expander can be to force the parent to
-        # allocate to us the whole space.
-        min_, nat = Gtk.HBox.do_get_preferred_width(self)
-        if gtk_version > (3, 19):
-            # if we get too large gtk calcs will overflow..
-            nat = max(nat, 2 ** 16)
-        return (min_, nat)
-
-
 class QueueExpander(Gtk.Expander):
 
     def __init__(self, library, player):
@@ -91,7 +77,7 @@ class QueueExpander(Gtk.Expander):
 
         add_css(self, ".ql-expanded title { margin-bottom: 5px; }")
 
-        outer = ExpandBoxHack()
+        outer = ExpanderTitleContainerHack()
 
         left = Gtk.HBox(spacing=12)
 
diff --git a/quodlibet/quodlibet/qltk/x.py b/quodlibet/quodlibet/qltk/x.py
index 0b74e8474..b5610bfd6 100644
--- a/quodlibet/quodlibet/qltk/x.py
+++ b/quodlibet/quodlibet/qltk/x.py
@@ -524,3 +524,22 @@ class HighlightToggleButton(Gtk.ToggleButton):
     def do_draw(self, context):
         self._update_provider()
         return Gtk.ToggleButton.do_draw(self, context)
+
+
+class ExpanderTitleContainerHack(Gtk.HBox):
+
+    def __init__(self, content=None, *args, **kwargs):
+        super(Gtk.HBox, self).__init__(*args, **kwargs)
+        if content:
+            self.add(content)
+
+    def do_get_preferred_width(self):
+        # Workaround for https://bugzilla.gnome.org/show_bug.cgi?id=765602
+        # set_label_fill() no longer works since 3.20. Fake a natural size
+        # which is larger than the expander can be to force the parent to
+        # allocate to us the whole space.
+        min_, nat = Gtk.HBox.do_get_preferred_width(self)
+        if gtk_version > (3, 19):
+            # if we get too large gtk calcs will overflow..
+            nat = max(nat, 2 ** 16)
+        return (min_, nat)
-- 
2.14.1

